[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions.

ipiszy opened the pull request:
This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* #107931
* #107901
* #107847
* __->__ #107802



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
Inductor config for cuda / cutlass, util functions.

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

kadeng commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,13 @@
+CUTLASS_PY_PATH = "cutlass/tools/library/scripts/"
+
+# Add CUTLASS Python scripts to Python search path.
+sys.path.insert(
```
A global modification of the sys.path is a very dangerous thing to do, and unexpected to happen on module load. Especially considering that namespace clashes are likely, considering what is below that path. See https://github.com/NVIDIA/cutlass/tree/main/tools/library/scripts - that location would override global packages named "library", "rt", "manifest", "generator" and even the top level __init__.py

While I see that there seems to be no easy workaround, I think this is too dangerous to merge into pytorch as is.

ipiszy replied:
Yes, this is a very valid concern.

1) I tried to update cutlass repo to use relative imports. However this breaks cutlass profiler so I had to revert this commit (https://github.com/NVIDIA/cutlass/compare/97b46e7549459f626f6b5c343acb6d75d089429d..12e5e4beb9dcc71af9708c80a5fb0168e599d555).
2) There are some discussions in the cutlass slack channel. Currently a preferred way is to add a new folder like "cutlass_scripts", create __init__.py under this folder and import submodules correctly.
3) It would take some time to make this change in cutlass. e.g. submit a PR, merge it in cutlass, sync it to Pytorch and internal, and update this piece of code. I don't want this PR to be blocked by the cutlass change.
4) I've also checked that under torch there are no existing libs naming like "library", "rt", "manifest", "generator". So it should be safe for now.
5) I'll work on the cutlass fix in parallel.

aakhundov replied:
Can we, maybe, (temporarily) get away with "private" imports? Something along the lines of importing the module with a helper function using `importlib` with the path to the module (possibly with `@cache`-ing). Would that work, without polluting the global `sys.modules` namespace?

ipiszy replied:
I'm not sure whether importlib works. I tried several tricks listed here but all failed: https://www.tutorialspoint.com/How-to-import-a-Python-module-given-the-full-path#:~:text=Using%20importlib.&text=of%20using%20importlib.-,util.,the%20module%20to%20be%20imported.
The reason is that these cutlass scripts import from other cutlass scripts, and during this process it still tries to find other cutlass scripts from path. It seems that the current way to import a module from a custom path doesn't work recursively.

jansel replied:
You might be able to do something similar to:
https://github.com/pytorch/pytorch/blob/1b2eac00cb6ebf7f51e8c7ed182b8915fa816f2b/test/inductor/test_compiled_autograd.py#L343-L348

It would be much better to put the cutlass stuff in a namespace rather than polluting the global one with these generic names.  There might be user code that uses these same names.

ipiszy replied:
Ah I see, I think `mock.patch` should work since it updates `sys.path`...
On the other hand I just submitted a PR to cutlass: https://github.com/NVIDIA/cutlass/pull/1072. If it could be accepted I'll just import from `third_party.cutlass.tools.library.scripts` directly.

kadeng replied:
I see a potential problem with the approach of patching sys.path and using importlib's SourceFileLoader: If I get it correctly, using an importlib.SourceFileLoader loader will not bypass importlib's caches, nor the global sys.modules. Even worse, it will populate these. See https://peps.python.org/pep-0302/

"If there is an existing module object named ‘fullname’ in sys.modules, the loader must use that existing module. (Otherwise, the reload() builtin will not work correctly.) If a module named ‘fullname’ does not exist in sys.modules, the loader must create a new module object and add it to sys.modules."

The SourceFileLoader would therefore still contaminate the global module scope ( and also read from it ).

So, having a clean package name would be preferable. If we don't have that, we might have to resolve to some intermediate hack, like copying all source files into a new directory (say "cutlass_scripts"), and replacing every line like "import library" to "import cutlass_scripts.library as library" (etc..) within these files. Could be done with a few simple regex replacements as we can easily test that it works on the few files we would have to apply this to. It could also be done in a packaging script, not necessarily at runtime.

ipiszy replied:
This PR was updated last week to not use importlib's SourceFileLoader. It's now copying cutlass directory and adding cutlass_ prefixes to all submodules.

kadeng commented on torch/_inductor/codegen/cuda/cuda_env.py:
```diff
@@ -0,0 +1,31 @@
+        if cuda_arch is None:
+            # Get Compute Capability of the first Visible device
+            major, minor = torch.cuda.get_device_capability(0)
+            cuda_arch = major * 10 + minor
```
this appears to be an int or some kind of number, not a str as in the type hint, right?

ipiszy replied:
Updated with correct type hints.

kadeng commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,158 @@
+
+    if torch_dtype0 == torch_dtype1:
+        return True
+    for matched_dtypes in _MATHED_DTYPES:
```
typo in _MATHED_DTYPES

kadeng commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,158 @@
+    Returns all possible alignments for a given torch dtype.
+    """
+
+    if torch_dtype in (torch.float16, torch.half, torch.bfloat16):
```
Is this correct? I would think that the alignment should be at least sizeof(dtype), which is 2 for 16 bit and 4 for 32 bit. Or do I misunderstand something here?

ipiszy replied:
This function returns the possible number of elements for a dtype which forms valid cutlass alignments.
On the cutlass side, the max alignment in bytes is 16 bytes, and min alignment in bytes is 2 bytes.

ipiszy replied:
Also updated comments, hope it looks clearer.

kadeng commented on torch/_inductor/config.py:
```diff
@@ -84,9 +84,10 @@
 # Triton: Triton templates defined in torch inductor.
+# CUTLASS: Cutlass templates and kernels.
 max_autotune_gemm_backends = os.environ.get(
     "TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS", "ATEN,TRITON"
```
this means CUTLASS backend is not enabled by default, right?

ipiszy replied:
Correct.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @kadeng !

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

aakhundov commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,15 @@
+# TODO(ipiszy): Update cutlass repo to export all submodules under "cutlass_scripts" to avoid
+# potential naming conflicts.
+sys.path.insert(
+    0,
```
Is it necessary to insert this at the beginning of the `sys.path`? If this is the case:

> I've also checked that under torch there are no existing libs naming like "library", "rt", "manifest", "generator".

then appending at the end should still work and won't cause as much trouble? Or do (some of) those names exist in one ore more of the existing paths?

aakhundov commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,15 @@
+import os
+import sys
+
+THIRD_PARTY_PATH = "../../../../third_party/"
```
Nit: This looks quite fragile in the face of possible file structure reorg. Maybe use `torch` module's path as the base?

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+    version = get_cuda_version()
+    if arch is None or version is None:
+        log.error(
+            "Cannot find cuda arch %s or cuda version %s. "
```
Nit: perhaps "Cannot detect ..."?

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+def dtype_match(torch_dtype: torch.dtype, cutlass_dtype: cutlass_lib.DataType) -> bool:
+    if torch_dtype == torch.float or torch_dtype == torch.float32:
+        return (
+            cutlass_dtype == cutlass_lib.DataType.f32
+            or cutlass_dtype == cutlass_lib.DataType.tf32
+        )
```
Nit: use `cutlass_dtype in {...}`? Same for `torch_dtype` in the conditions.

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+def get_alignments(torch_dtype: torch.dtype) -> List[int]:
+    """
+    Returns all possible number of elements for a dtype which forms valid cutlass alignments.
```
This sentence is not clear. I understand what you mean, but that's because I have context, I believe. Maybe this instead:

> Returns all possible valid CUTLASS alignments in terms of the number of elements for a given dtype.

Also nit: let's try to keep `CUTLASS` upper-case in the docs?

aakhundov commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+        raise NotImplementedError(f"unsupported {torch_dtype=} for alignments")
+
+
+def get_alignment(inductor_layout: Layout) -> int:
```
Nit: rename this to `get_max_alignment`? Because that's what the docstring says.

aakhundov commented on torch/_inductor/config.py:
```diff
@@ -400,6 +401,33 @@ class triton:
+    version = None
+
+    # Optimization level for the host compiler.
+    compile_opt_level = "-O1"
```
`-O1` looks a bit strange here as a default. Shouldn't we have either `-O0` (for no optimization) or `-O3` (for max. optimization) by default?

jansel commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,15 @@
+import os
+import sys
+
+THIRD_PARTY_PATH = "../../../../third_party/"
+CUTLASS_PY_PATH = "cutlass/tools/library/scripts/"
```
I don't think this will work with `pip install torch` since the `third_party` folder is not part of our python package.  Will need to copy the needed stuff into our packaged torch/headers folder.

ipiszy replied:
`pip install torch` won't work. Currently only local development works.
I think @atalman has some plans to export cutlass files to torch package correctly (there is also an issue tracked in https://github.com/pytorch/pytorch/issues/106991). 

On the other hand, I just noticed that some torch tests are also based on torch package, which blocks the latter PRs in the stack. Will work with Andrey to resolve this issue.

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+from typing import Any, List
+
+# Import cutlass python scripts.
+import generator as cutlass_generator  # type: ignore[import]
+import library as cutlass_lib  # type: ignore[import]
+import manifest as cutlass_manifest  # type: ignore[import]
```
These are very generic names that could be used in user code.  Given how many users of PyTorch there are I am sure someone has a module named `library` in some closed source project.

We should put these in some sort of `cutlass_foo.*` namespace.

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+log = logging.getLogger(__name__)
+
+
+class Args:
```
More descriptive name

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+    CUTLASS args used to initialize a CUTLASS Manifest.
+    """
+
+    def __init__(self, cuda_arch, cuda_version):
+        self.operations = "all"
+        self.build_dir = ""
+        self.curr_build_dir = ""
+        self.generator_target = ""
+        self.architectures = cuda_arch
+        self._normalize_cuda_arch()
+        self.kernels = "all"
+        self.ignore_kernels = ""
+        self.cuda_version = cuda_version
+        self.kernel_filter_file = None
+        self.selected_kernel_list = None
+        self.interface_dir = None
+        self.filter_by_cc = True
+        self.disable_full_archs_compilation = False
```
suggest using dataclass with `_post_init`

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+    if torch_dtype0 == torch_dtype1:
+        return True
+    for matched_dtypes in _MATCHED_DTYPES:
+        if torch_dtype0 in matched_dtypes and torch_dtype1 in matched_dtypes:
```
We should be able to simplify this to not use a loop.

jansel replied:
Is this even needed?
```
>>> import torch
>>> torch.float == torch.float32
True
>>> torch.half == torch.float16
True
>>> 
```

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+
+
+def dtype_match(torch_dtype: torch.dtype, cutlass_dtype: cutlass_lib.DataType) -> bool:
+    if torch_dtype == torch.float or torch_dtype == torch.float32:
```
```
>>> torch.float == torch.float32
True
```
can conbine the two tests?   (same below)

ipiszy commented:
The cutlass team is discussing how to reorganize the python scripts under the cutlass repo. Will wait for cutlass team's update before I start to address comments.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@jansel @aakhundov Now I copy all cutlass python scripts to a temporary folder and add "cutlass_" prefix to all file names. I checked `sys.modules` to make sure that it's not polluted by general names like "library". ptal.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,77 @@
+HAS_CUTLASS = False
+
+_CUTLASS_PY_FULL_PATH = os.path.join(
+    inductor_cuda_config.cutlass_dir, "tools/library/scripts"
```
config values can change after pytorch is imported, so we shouldn't do this in global scope

ipiszy replied:
Yeah makes sense, will fix this.

jansel commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,77 @@
+# This is a temporary hack to avoid CUTLASS module naming conflicts.
+# TODO(ipiszy): remove this hack when CUTLASS solves Python scripts packaging structure issues.
+if os.path.isdir(_CUTLASS_PY_FULL_PATH):
+    cutlass_file_names = [
+        file_name
+        for file_name in os.listdir(_CUTLASS_PY_FULL_PATH)
+        if file_name.endswith(".py")
+    ]
+    cutlass_module_names = [file_name[:-3] for file_name in cutlass_file_names]
+    if not os.path.isdir(_TMP_CUTLASS_PY_FULL_PATH):
+        os.mkdir(_TMP_CUTLASS_PY_FULL_PATH)
+    for file_name in cutlass_file_names:
+        _gen_cutlass_file(file_name, cutlass_module_names)
+    sys.path.append(_TMP_CUTLASS_PY_FULL_PATH)
+    try:
+        import cutlass_generator  # type: ignore[import]
+        import cutlass_library  # type: ignore[import]
+        import cutlass_manifest  # type: ignore[import]
+
+        HAS_CUTLASS = True
+    except ImportError as e:
+        log.warning(
+            "Failed to import CUTLASS packages: %s, ignoring the CUTLASS backend.",
+            str(e),
+        )
```
Let's move this to be done lazily with a `@functools.lru_cache(None)` decorator, rather than an import time.

jansel commented on torch/_inductor/codegen/cuda/__init__.py:
```diff
@@ -0,0 +1,77 @@
+            str(e),
+        )
+else:
+    log.warning(
+        "Failed to import CUTLASS packages: CUTLASS repo does not exist: %s",
+        _CUTLASS_PY_FULL_PATH,
+    )
```
Will this warning fire for all `pip`/`conda` users?  Maybe we should get it on the cutlass path existing, or being manually set?

ipiszy replied:
Updated this to log.debug here, and only do warning when CUTLASS option is explicitly set in PR 5.

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+    for dtype in input_torch_dtypes[1:]:
+        if torch_dtype != dtype:
+            raise RuntimeError(f"Unmatched input dtypes: {torch_dtype=}, {dtype=}")
+    if torch_dtype in {torch.float16, torch.half}:
```
arent float16 and half the same type? `== torch.float16` might be the same thing?

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,170 @@
+            return torch_dtype
+        else:
+            return torch.float
+    if torch_dtype in {torch.bfloat16, torch.float, torch.float32}:
```
Shoudl we use allow_bf16_reduced_precision_reduction for torch.bfloat16?

ipiszy replied:
CUTLASS doesn't support using bf16 as the accumulation type. Tensor core only supports fp32 or fp16 accumulation.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
Hi @jansel , I've addressed comments, ptal, thanks!

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cuda/cutlass_utils.py:
```diff
@@ -0,0 +1,252 @@
+        inductor_cuda_config.cutlass_dir, "tools/library/scripts"
+    )
+    tmp_cutlass_py_full_path = os.path.abspath(
+        os.path.join(tempfile.gettempdir(), "torch_cutlass_script")
```
This might cause issues with multiple users on the same system.  One user might not have permission to import something created by another user.

Let's put it inside https://github.com/pytorch/pytorch/blob/fa542cc4bb449cd05c81a17bd0489e46dcb4f1ba/torch/_inductor/codecache.py#L96-L101

ipiszy replied:
Makes sense, will put it inside the cache_dir in codecache.py.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@pytorchbot merge

ipiszy commented:
@pytorchbot label "topic: not user facing"

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 1: Inductor config for cuda / cutlass, util functions."

This is the step 1 to add cutlass as an alternative inductor backend.
Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]
